server/src/main/scala/org/http4s/server/ServerApp.scala

```diff
@@ -0,0 +1,96 @@
+package org.http4s
+package server
+
+import java.net.ServerSocket
+import java.util.concurrent.atomic.AtomicReference
+import java.util.concurrent.{ CountDownLatch, RejectedExecutionException }
+import java.util.concurrent.atomic.AtomicBoolean
+import java.util.concurrent.{ Executors, ThreadFactory }
+import scala.annotation.tailrec
+import scalaz.concurrent.Task
+import org.http4s.util.threads
+
+/**
+ * Apps extending the server app trait get a graceful shutdown.  The
+ * 
+ */ 
+trait ServerApp {
+  private[this] val logger = org.log4s.getLogger
+
+  /** Return a server to run */
+  def server(args: List[String]): Task[Server]
+
+  /** Return a task to shutdown the application.
+   * 
+   *  This task is run as a JVM shutdown hook, or when
+   *  [[requestShutdown()]] is explicitly called.
+   *  
+   *  The default implementation shuts down the server, and waits for
+   *  it to finish.  Other resources may shutdown by flatMapping this
+   *  task.
+   */
+  def shutdown(server: Server): Task[Unit] =
+    server.shutdown
+
+  private sealed trait LifeCycle
+  private case object Init extends LifeCycle
```

Comment:
I think this one is actually not really necessary.  We don't ever distinguish Init and Starting.